sigmf/archive.py

```diff
@@ -0,0 +1,140 @@
+                   if specified will be used to determine the directory and
+                   file names within the archive. `fileobj` won't be closed.
+                   For example: if `name` == "archive1" and fileobj is given,
+                   a tar archive will be written to fileobj with the
+                   following structure:
+                       - archive1/
+                         - archive1.sigmf-meta
+                         - archive1.sigmf-data
+
+    """
+    def __init__(self, sigmffile, name=None, fileobj=None):
+        self.sigmffile = sigmffile
+        self.name = name
+        self.fileobj = fileobj
+
+        self._validate()
+
+        archive_name = self._get_name()
+        sigmf_fd = self._open()
+        sigmf_archive = tarfile.TarFile(mode="w",
+                                        fileobj=sigmf_fd)
+        tmpdir = tempfile.mkdtemp()
+        sigmf_md_filename = archive_name + SIGMF_METADATA_EXT
+        sigmf_md_path = os.path.join(tmpdir, sigmf_md_filename)
+        sigmf_data_filename = archive_name + SIGMF_DATASET_EXT
+        sigmf_data_path = os.path.join(tmpdir, sigmf_data_filename)
+
+        with open(sigmf_md_path, "w") as mdfile:
+            self.sigmffile.dump(mdfile, pretty=True)
+
+        shutil.copy(self.sigmffile.data_file, sigmf_data_path)
+
+        def chmod(tarinfo):
+            if tarinfo.isdir():
+                tarinfo.mode = 0o755  # dwrxw-rw-r
+            else:
+                tarinfo.mode = 0o644  # -wr-r--r--
+            return tarinfo
+
+        sigmf_archive.add(tmpdir, arcname=archive_name, filter=chmod)
+        sigmf_archive.close()
+        if not fileobj:
+            sigmf_fd.close()
+
+        shutil.rmtree(tmpdir)
+
+        self.path = sigmf_archive.name
+
+    def _validate(self):
+        name = self.name
+        if name and "." in name and not name.endswith(SIGMF_ARCHIVE_EXT):
+            apparent_ext = os.path.splitext(name)[-1]
+            err = "extension {} != {}".format(apparent_ext, SIGMF_ARCHIVE_EXT)
+            raise error.SigMFFileError(err)
+
+        valid_md = self.sigmffile.validate()
+        if not valid_md:
+            err = "invalid metadata - {!s}"
+            raise error.SigMFValidationError(err.format(valid_md))
+
+        if not self.sigmffile.data_file:
+            err = "no data file - use `set_data_file`"
+            raise error.SigMFFileError(err)
+
+    def _get_name(self):
+        if self.fileobj and not self.name:
+            filename = os.path.split(self.fileobj.name)[-1]
+        else:
+            filename = os.path.split(self.name)[-1]
+
+        archive_name = os.path.splitext(filename)[0]
+        return archive_name
+
+    def _open(self):
+        if self.fileobj:
+            sigmf_fd = self.fileobj
+        else:
+            has_ext = self.name.endswith(SIGMF_ARCHIVE_EXT)
+            name = self.name if has_ext else self.name + SIGMF_ARCHIVE_EXT
+            sigmf_fd = open(name, "wb")
```

Comment:
This could definitely throw an error, so it might be nice to catch it. If this were a command-line utility, I would definitely catch it and print out the error and then exit, but since it's designed to be imported by another program, the polite thing to do is let exceptions raise (rise?). We could catch it and wrap it in one of the SigMFError exception classes I created, but the few cases where I used those I didn't use them to catch and wrap existing exceptions, I just didn't want to raise something generic like RuntimeError. My personal feeling is that just letting the original error raise is a bit cleaner than wrapping it and re-raising. Especially in python3 where you will get the traceback chain with the original error anyways!

```python
>>> t = tempfile.mkstemp()
>>> t
(3, '/tmp/tmpl4smirjf')
>>> f = open(t[1], mode="r")
>>> f.writable()
False
>>> f.write("stuff")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
io.UnsupportedOperation: not writable
>>> try:
...   f.write("stuff")
... except Exception as err:
...   raise RuntimeError(err)  # pretending this is SigMFFileError or something
... 
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
io.UnsupportedOperation: not writable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 4, in <module>
RuntimeError: not writable
>>> 
```

I dunno, considering we will need to raise anyways, what do you think?